From 95e0d21105421e45ce875d3db1dc95924031af97 Mon Sep 17 00:00:00 2001
From: Dale Curtis <dalecurtis@chromium.org>
Date: Wed, 23 Oct 2024 02:20:37 +0000
Subject: [PATCH] Bump maximum video frame size by one bit.

This increases the maximum canvas size from 268,435,456 or roughly
16384x16384 to 536,870,911 or roughly 23170x23170. This increase
still allows us to represent up to 8 bytes per pixel (which matches
our largest pixel formats) without overflowing a uint32_t.

This also updates the file to use inline constexpr per the
presubmit check and adds tests for these allocation limits.

Fixed: 374807001
Change-Id: I7eef8b488fa758bfcd0202917469d585724e23ba
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5955549
Commit-Queue: Eugene Zemtsov <eugene@chromium.org>
Auto-Submit: Dale Curtis <dalecurtis@chromium.org>
Reviewed-by: Eugene Zemtsov <eugene@chromium.org>
Commit-Queue: Dale Curtis <dalecurtis@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1372466}
---

--- a/media/base/limits.h
+++ b/media/base/limits.h
@@ -10,18 +10,21 @@
 #include "build/build_config.h"
 #include "media/media_buildflags.h"
 
-namespace media {
-
-namespace limits {
+namespace media::limits {
 
 // Maximum possible dimension (width or height) for any video.
-constexpr int kMaxDimension = (1 << 15) - 1;  // 32767
+inline constexpr int kMaxDimension = (1 << 15) - 1;  // 32767
 
 // Maximum possible canvas size (width multiplied by height) for any video.
-constexpr int kMaxCanvas = (1 << (14 * 2));  // 16384 x 16384
+// This lets us represent up to 8 bytes per pixel without overflowing uint32_t.
+//
+// The actual allocation limit on 32-bit platforms is ~6 bytes per pixel since
+// we're unable to allocate more than 2gb, so creating frames above this limit
+// will fail (gracefully) if allocation is actually attempted.
+inline constexpr int kMaxCanvas = (1 << 29) - 1;  // 23170 x 23170
 
 // Total number of video frames which are populating in the pipeline.
-constexpr int kMaxVideoFrames = 4;
+inline constexpr int kMaxVideoFrames = 4;
 
 // The following limits are used by AudioParameters::IsValid().
 //
@@ -35,58 +38,56 @@
 //
 // kMaxSampleRate should be updated with
 // blink::audio_utilities::MaxAudioBufferSampleRate()
-constexpr int kMaxSampleRate = 768000;
-constexpr int kMinSampleRate = 3000;
-constexpr int kMaxChannels = 32;
-constexpr int kMaxBytesPerSample = 4;
-constexpr int kMaxBitsPerSample = kMaxBytesPerSample * 8;
-constexpr int kMaxSamplesPerPacket = kMaxSampleRate;
-constexpr int kMaxPacketSizeInBytes =
+inline constexpr int kMaxSampleRate = 768000;
+inline constexpr int kMinSampleRate = 3000;
+inline constexpr int kMaxChannels = 32;
+inline constexpr int kMaxBytesPerSample = 4;
+inline constexpr int kMaxBitsPerSample = kMaxBytesPerSample * 8;
+inline constexpr int kMaxSamplesPerPacket = kMaxSampleRate;
+inline constexpr int kMaxPacketSizeInBytes =
     kMaxBytesPerSample * kMaxChannels * kMaxSamplesPerPacket;
 
 // This limit is used by ParamTraits<VideoCaptureParams>.
-constexpr int kMaxFramesPerSecond = 1000;
+inline constexpr int kMaxFramesPerSecond = 1000;
 
 // The minimum elapsed amount of time (in seconds) for a playback to be
 // considered as having active engagement.
-constexpr int kMinimumElapsedWatchTimeSecs = 7;
+inline constexpr int kMinimumElapsedWatchTimeSecs = 7;
 
 // Maximum lengths for various EME API parameters. These are checks to
 // prevent unnecessarily large parameters from being passed around, and the
 // lengths are somewhat arbitrary as the EME spec doesn't specify any limits.
-constexpr int kMinCertificateLength = 128;
-constexpr int kMaxCertificateLength = 16 * 1024;
-constexpr int kMaxSessionIdLength = 512;
-constexpr int kMinKeyIdLength = 1;
-constexpr int kMaxKeyIdLength = 512;
-constexpr int kMaxKeyIds = 128;
-constexpr int kMaxInitDataLength = 64 * 1024;         // 64 KB
-constexpr int kMaxSessionResponseLength = 64 * 1024;  // 64 KB
-constexpr int kMaxKeySystemLength = 256;
+inline constexpr int kMinCertificateLength = 128;
+inline constexpr int kMaxCertificateLength = 16 * 1024;
+inline constexpr int kMaxSessionIdLength = 512;
+inline constexpr int kMinKeyIdLength = 1;
+inline constexpr int kMaxKeyIdLength = 512;
+inline constexpr int kMaxKeyIds = 128;
+inline constexpr int kMaxInitDataLength = 64 * 1024;         // 64 KB
+inline constexpr int kMaxSessionResponseLength = 64 * 1024;  // 64 KB
+inline constexpr int kMaxKeySystemLength = 256;
 
 // Minimum and maximum buffer sizes for certain audio platforms.
 #if BUILDFLAG(IS_MAC)
-constexpr int kMinAudioBufferSize = 128;
-constexpr int kMaxAudioBufferSize = 4096;
+inline constexpr int kMinAudioBufferSize = 128;
+inline constexpr int kMaxAudioBufferSize = 4096;
 #elif BUILDFLAG(USE_CRAS)
 // Though CRAS has different per-board defaults, allow explicitly requesting
 // this buffer size on any board.
-constexpr int kMinAudioBufferSize = 256;
-constexpr int kMaxAudioBufferSize = 8192;
+inline constexpr int kMinAudioBufferSize = 256;
+inline constexpr int kMaxAudioBufferSize = 8192;
 #endif
 
 // Maximum buffer size supported by Web Audio.
-constexpr int kMaxWebAudioBufferSize = 8192;
+inline constexpr int kMaxWebAudioBufferSize = 8192;
 
 // Bounds for the number of threads used for software video decoding.
-constexpr int kMinVideoDecodeThreads = 2;
-constexpr int kMaxVideoDecodeThreads =
+inline constexpr int kMinVideoDecodeThreads = 2;
+inline constexpr int kMaxVideoDecodeThreads =
     16;  // Matches ffmpeg's MAX_AUTO_THREADS. Higher values can result in
          // immediate out of memory errors for high resolution content. See
          // https://crbug.com/893984
 
-}  // namespace limits
-
-}  // namespace media
+}  // namespace media::limits
 
 #endif  // MEDIA_BASE_LIMITS_H_
--- a/media/base/video_frame.cc
+++ b/media/base/video_frame.cc
@@ -1063,10 +1063,14 @@
 // static
 size_t VideoFrame::AllocationSize(VideoPixelFormat format,
                                   const gfx::Size& coded_size) {
-  size_t total = 0;
-  for (size_t i = 0; i < NumPlanes(format); ++i)
-    total += PlaneSize(format, i, coded_size).GetArea();
-  return total;
+  base::CheckedNumeric<size_t> total = 0;
+  for (size_t i = 0; i < NumPlanes(format); ++i) {
+    total += PlaneSize(format, i, coded_size).Area64();
+  }
+
+  // If the line below is crashing, someone has likely added a pixel format with
+  // more than 8 bytes per pixel.
+  return total.ValueOrDie();
 }
 
 // static
--- a/media/base/video_frame_unittest.cc
+++ b/media/base/video_frame_unittest.cc
@@ -25,6 +25,7 @@
 #include "build/build_config.h"
 #include "gpu/command_buffer/common/mailbox_holder.h"
 #include "media/base/color_plane_layout.h"
+#include "media/base/limits.h"
 #include "media/base/simple_sync_token_client.h"
 #include "media/video/fake_gpu_memory_buffer.h"
 #include "testing/gtest/include/gtest/gtest.h"
@@ -853,6 +854,35 @@
   }
 }
 
+// Test ensures we don't overflow on 32-bit platforms.
+TEST(VideoFrame, NoFrameSizeExceedsUint32) {
+  const int max_dimension = std::sqrt(limits::kMaxCanvas);
+  const auto max_size = gfx::Size(max_dimension, max_dimension);
+  for (unsigned int i = 1u; i <= PIXEL_FORMAT_MAX; ++i) {
+    // Deprecated pixel formats.
+    if (i == 13 || i == 15 || i == 25) {
+      continue;
+    }
+
+    const auto format = static_cast<VideoPixelFormat>(i);
+
+    ASSERT_TRUE(
+        VideoFrame::IsValidConfig(format, VideoFrame::STORAGE_UNOWNED_MEMORY,
+                                  max_size, gfx::Rect(max_size), max_size));
+
+    base::CheckedNumeric<uint32_t> allocation_size =
+        VideoFrame::AllocationSize(format, max_size);
+    ASSERT_TRUE(allocation_size.IsValid());
+
+    // Allocations above 2gb on 32-bit platforms should fail gracefully.
+#ifdef ARCH_CPU_32_BITS
+    if (allocation_size.ValueOrDie() >= std::numeric_limits<int32_t>::max()) {
+      ASSERT_FALSE(VideoFrame::CreateBlackFrame(max_size));
+    }
+#endif
+  }
+}
+
 TEST(VideoFrame, WrapExternalDataWithInvalidLayout) {
   auto coded_size = gfx::Size(320, 180);
 
--- a/media/base/video_types.h
+++ b/media/base/video_types.h
@@ -21,6 +21,9 @@
 // When a VideoFrame is backed by native textures, VideoPixelFormat describes
 // how those textures should be sampled and combined to produce the final
 // pixels.
+//
+// WARNING: Do not add pixel formats with more than 8 bytes per pixel or the
+// total size of the frame can overflow on 32-bit platforms.
 enum VideoPixelFormat {
   PIXEL_FORMAT_UNKNOWN = 0,  // Unknown or unspecified format value.
   PIXEL_FORMAT_I420 =
--- a/third_party/blink/web_tests/external/wpt/webcodecs/video-encoder-config.https.any.js
+++ b/third_party/blink/web_tests/external/wpt/webcodecs/video-encoder-config.https.any.js
@@ -135,8 +135,8 @@
     config: {
       codec: 'vp8',
       hardwareAcceleration: 'prefer-hardware',
-      width: 20000,
-      height: 20000,
+      width: 30000,
+      height: 30000,
       bitrate: 1,
       framerate: 240,
     }
--- a/third_party/blink/web_tests/fast/mediacapturefromelement/CanvasCaptureMediaStream-canvas-initially-too-large.html
+++ b/third_party/blink/web_tests/fast/mediacapturefromelement/CanvasCaptureMediaStream-canvas-initially-too-large.html
@@ -16,8 +16,8 @@
   canvas.height = 32768;
   assert_throws_dom("NotSupportedError", function() { canvas.captureStream() },
                     "video frame height too large");
-  canvas.width = 16385;
-  canvas.height = 16384;
+  canvas.width = 30000;
+  canvas.height = 30000;
   assert_throws_dom("NotSupportedError", function() { canvas.captureStream() },
                     "video frame area too large");
 });
